tests/test_nblist.py

```diff
@@ -376,7 +380,148 @@ def test_nblist_max_interactions(block_size, tiles):
+    host_config: Optional[HostConfig] = None
+
+    mol_a, mol_b, core = get_hif2a_ligand_pair_single_topology()
+    if host_name == "complex":
+        with resources.path("timemachine.testsystems.data", "hif2a_nowater_min.pdb") as protein_path:
+            host_sys, host_conf, box, _, num_water_atoms = build_protein_system(
+                str(protein_path), forcefield.protein_ff, forcefield.water_ff, mols=[mol_a, mol_b]
+            )
+            box += np.diag([0.1, 0.1, 0.1])  # remove any possible clashes
+        host_config = HostConfig(host_sys, host_conf, box, num_water_atoms)
+    elif host_name == "solvent":
+        host_sys, host_conf, box, _ = build_water_system(4.0, forcefield.water_ff, mols=[mol_a, mol_b])
+        box += np.diag([0.1, 0.1, 0.1])  # remove any possible clashes
+        host_config = HostConfig(host_sys, host_conf, box, host_conf.shape[0])
+    else:
+        assert 0, "Invalid host name"
+
+    st = SingleTopology(mol_a, mol_b, core, forcefield)
+    host = setup_optimized_host(st, host_config)
+
+    lambda_grid = np.array([0.0])
+
+    initial_state = setup_initial_states(st, host, DEFAULT_TEMP, lambda_grid, seed=2024, min_cutoff=None)[0]
+
+    return st, host, host_name, initial_state
+
+
+def compute_mean_tile_density(
+    nblist, frame: NDArray, box: NDArray, cutoff: float, column_block: int = 32, row_block: int = 32
+) -> float:
+    assert column_block == 32
+    assert column_block == row_block
+    N = frame.shape[0]
+    box_diag = np.diagonal(box)
+    assert nblist.get_num_row_idxs() == N
+    ixn_list = nblist.get_nblist(frame, box, cutoff)
+    tile_densities = []
+    num_blocks = (N + row_block - 1) // row_block
+    ixns_per_tile = min(N, column_block) * min(N, row_block)
+
+    row_block_offset = 0
+    # Each ixn_block is the column atoms in the tile that interact with at least one
+    # row block atom
+    for i, ixn_block in enumerate(ixn_list):
+        row_coords = frame[row_block_offset : row_block_offset + row_block]
+        row_coords = np.expand_dims(row_coords, axis=0)
+        column_ixn_atoms = np.array(ixn_block)
+        column_block_offset = 0
+        for _ in range(num_blocks - i):
+            # Find the interactions that would make up a tile, if the tile is empty
+            # don't compute density.
+            tile_column_ixns = column_ixn_atoms[
+                (column_ixn_atoms >= column_block_offset) & (column_ixn_atoms < (column_block_offset + column_block))
+            ]
+            column_block_offset += column_block
+            if tile_column_ixns.size == 0:
+                continue
+            col_coords = np.expand_dims(frame[tile_column_ixns], axis=1)
+            deltas = image_coords(row_coords - col_coords, box_diag)
+            dij = np.linalg.norm(deltas, axis=-1)
+            ixns = np.sum(dij < cutoff)
+            tile_densities.append(ixns / ixns_per_tile)
+        row_block_offset += row_block
+    return float(np.mean([tile_densities]))
```

Comment:
small observation: since `ixns_per_tile` is not constant, `mean(tile_densities)` will differ slightly from `sum(ixns) / sum(ixns_per_tile)`. The computed quantity does seem like a more performance-relevant number than `sum(ixns) / sum(ixns_per_tile)`.

Maybe a few other performance-relevant numbers to monitor might include:
* how many warp-synchronous loop iterations will be executed, vs. the bare minimum number required. The idealized minimum number might be something like `ceil(sum(ixns) / 32)`, but the number executed might be something like the number of rows (or is it columns?) that have at least 1 `dij < cutoff`, summed over all tiles.
* how many 32x32 tiles are produced, vs. the bare minimum number of 32x32 tiles
* fraction of atoms per tile row/column that have interactions (there was a comment in [[Eastman, Pande, 2009]](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2841709/) about having two code paths depending on whether 12 or fewer of the 32 atoms in a block have interactions)

Reply:
Would it be informative to plot the full distribution of tile densities? If so, could this function return the list of densities? (the mean could still be computed where we print the summary)